### ISSUE TITLE
Support KSA per Adapter instance

### ISSUE THREAD

**Issue by automaticgiant-ml** [triggermesh/triggermesh#1299]
Right now we have support for passing annotations from the Adapter object into the Kubernetes Service Account that works for Workload Identity triggermesh/triggermesh#1223.
But we run into complications/mismatch in our architecture and security posture when using more than one adapter in a Namespace.

- We have namespaces where there are multiple applications and pubsub subscriptions. These can be to topics in other projects, and our pubsub system (gcp pubsub-based) is intended to give granular access to these topics/subscriptions.
- The other issue is that we don't want one Application's resources to depend on resources in another application - we want the Service Account created in one place, but we don't want App A to only work if App B is deployed because they share a service account.

The current implementation of the GoogleCloudPubSubSource shares a KSA between all adapter deployments in a namespace, and this is not compatible with annotating multiple GoogleCloudPubSubSources with different GSA for Workload Identity.

Two possible/good options are:
- add the ServiceAccount name for the resulting KSA as a CRD parameter
- have the controller create one KSA per adapter instance and copy the name of the GoogleCloudPubSubSource object

@sebgoa @tzununbekov @snuggie12
I am unclear if we can commit (or have suitable) developer resources to contribute a PR, so please let us know what y'all see ahead for this feature request. Thanks so much!

**Comment by sebgoa** [comment]
@automaticgiant-ml  if you can/want to submit a PR we will certainly review it and work with you to get it merged.

it would be good to talk in real-time to refine the feature and see if we can schedule it. Feel to reach me sebgoa@triggermesh.com

**Comment by antoineco** [comment]
tip: This sounds like this would have worked as intended if #1223 had been implemented using the existing `ServiceAccountProvider` interface instead of adding a new mechanism which alters a shared serviceAccount:

https://github.com/triggermesh/triggermesh/blob/04e295f3e1a41cad5e81c1c5bcf86bae05ad9a08/pkg/apis/sources/v1alpha1/awssqs_lifecycle.go#L74-L88

By simply implementing this interface, the reconciler creates one serviceAccount per source instance, which sounds like what you're requesting.

**Comment by antoineco** [comment]
tip: This sounds like this would have worked as intended if #1223 had been implemented using the existing [`ServiceAccountProvider` interface](https://github.com/triggermesh/triggermesh/blob/975fc8f3d8508db7c3452cddbaf28ec78efda55e/pkg/apis/common/v1alpha1/object_interfaces.go#L84-L89) instead of adding a new mechanism based on Context injection which alters the shared serviceAccount.

By simply implementing this interface, the reconciler creates one serviceAccount per source instance, which sounds like what you're requesting.

Example:

```go
// WantsOwnServiceAccount implements ServiceAccountProvider.
func (s *GoogleCloudPubSubSource) WantsOwnServiceAccount() bool {
	return s.Spec.Auth.GcpServiceAccount != nil
}

// ServiceAccountOptions implements ServiceAccountProvider.
func (s *GoogleCloudPubSubSource) ServiceAccountOptions() []resource.ServiceAccountOption {
	var saOpts []resource.ServiceAccountOption

	if gcpSA := s.Spec.Auth.GcpServiceAccount; gcpSA != nil {
		saOpts = append(saOpts, func(sa *corev1.ServiceAccount) {
			metav1.SetMetaDataAnnotation(&sa.ObjectMeta, "iam.gke.io/gcp-service-account", *gcpSA)
		})
	}

	return saOpts
}
```

**Comment by sebgoa** [comment]
@automaticgiant-ml we have a PR at work to fix this, care to comment https://github.com/triggermesh/triggermesh/pull/1316

**Comment by sebgoa** [comment]
this is released on v1.24, I am closing the ticket. Let us know how that works for you, thanks for the issue.
